File: Course-Project/app/src/main/java/com/vladislavmyasnikov/courseproject/ui/courses/AcademicPerformanceFragment.java
package com.vladislavmyasnikov.courseproject.ui.courses;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;

import android.os.Handler;
import android.os.Looper;
import android.os.Message;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.vladislavmyasnikov.courseproject.R;
import com.vladislavmyasnikov.courseproject.utilities.DataUpdater;
import com.vladislavmyasnikov.courseproject.ui.main.interfaces.OnRefreshLayoutListener;
import com.vladislavmyasnikov.courseproject.ui.components.UserView;

public class AcademicPerformanceFragment extends Fragment {

    private static final int CURRENT_HARDCODED_NUMBER_OF_USER_ICONS = 2;

    private OnRefreshLayoutListener mRefreshLayoutListener;
    private UserView mUser1View;
    private UserView mUser2View;
    private Handler mHandler = new Handler(Looper.getMainLooper()) {
        @Override
        public void handleMessage(Message msg) {
            super.handleMessage(msg);
            Bundle data = msg.getData();
            int[] points = data.getIntArray(DataUpdater.UPDATED_POINTS_DATA);
            if (points != null && mRefreshLayoutListener != null) {
                mUser1View.setBadgeCount(points[0]);
                mUser2View.setBadgeCount(points[1]);
                mRefreshLayoutListener.stopRefreshing();
            }
        }
    };

    private View.OnClickListener mOnTitleListener = new View.OnClickListener() {
        @Override
        public void onClick(View v) {
            Intent intent = new Intent(getContext(), UsersListActivity.class);
            startActivity(intent);
        }
    };

    @Override
    public void onAttach(@NonNull Context context) {
        super.onAttach(context);
        if (getParentFragment() instanceof OnRefreshLayoutListener) {
            mRefreshLayoutListener = (OnRefreshLayoutListener) getParentFragment();
        } else {
            throw new IllegalStateException(getParentFragment() + " must implement OnRefreshLayoutListener");
        }
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_academic_performance, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, Bundle savedInstanceState) {
        mUser1View = view.findViewById(R.id.user_1);
        mUser2View = view.findViewById(R.id.user_2);
        view.findViewById(R.id.title).setOnClickListener(mOnTitleListener);
    }

    @Override
    public void onDetach() {
        super.onDetach();
        mRefreshLayoutListener = null;
    }

    public void updateBadges() {
        DataUpdater.newInstance(mHandler, CURRENT_HARDCODED_NUMBER_OF_USER_ICONS).start();
    }

    public static AcademicPerformanceFragment newInstance() {
        return new AcademicPerformanceFragment();
    }
}

File: Course-Project/app/src/main/java/com/vladislavmyasnikov/courseproject/ui/courses/AcademicPerformanceFragment.kt
package com.vladislavmyasnikov.courseproject.ui.courses

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProviders
import androidx.recyclerview.widget.RecyclerView
import com.vladislavmyasnikov.courseproject.R
import com.vladislavmyasnikov.courseproject.di.components.DaggerAcademicPerformanceFragmentInjector
import com.vladislavmyasnikov.courseproject.di.modules.ContextModule
import com.vladislavmyasnikov.courseproject.ui.adapters.StudentAdapter
import com.vladislavmyasnikov.courseproject.ui.main.App
import com.vladislavmyasnikov.courseproject.ui.main.interfaces.UpdateStartListener
import com.vladislavmyasnikov.courseproject.ui.main.interfaces.UpdateStopListener
import com.vladislavmyasnikov.courseproject.ui.viewmodels.StudentListViewModel
import com.vladislavmyasnikov.courseproject.ui.viewmodels.StudentListViewModelFactory
import javax.inject.Inject

class AcademicPerformanceFragment : Fragment(), UpdateStartListener {

    @Inject
    lateinit var studentListViewModelFactory: StudentListViewModelFactory

    @Inject
    lateinit var mAdapter: StudentAdapter

    private lateinit var mStudentListViewModel: StudentListViewModel

    private var mUpdateStopListener: UpdateStopListener? = null

    private val mOnTitleListener = View.OnClickListener {
        val intent = Intent(context, StudentListActivity::class.java)
        startActivity(intent)
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)

        mUpdateStopListener = when {
            context is UpdateStopListener -> context
            parentFragment is UpdateStopListener -> parentFragment as UpdateStopListener
            else -> throw IllegalStateException("Context or parent fragment must implement UpdateStopListener")
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        return inflater.inflate(R.layout.fragment_academic_performance, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        view.findViewById<View>(R.id.title).setOnClickListener(mOnTitleListener)

        val injector = DaggerAcademicPerformanceFragmentInjector.builder().appComponent(App.appComponent).contextModule(ContextModule(activity!!)).build()
        injector.injectAcademicPerformanceFragment(this)
        mStudentListViewModel = ViewModelProviders.of(this, studentListViewModelFactory).get(StudentListViewModel::class.java)

        val recyclerView = view.findViewById<RecyclerView>(R.id.recycler_view)
        mAdapter.viewType = StudentAdapter.ViewType.COMPACT_VIEW
        recyclerView.adapter = mAdapter
    }

    override fun startUpdate() {
        mStudentListViewModel.fetchStudents()
    }

    override fun onDetach() {
        super.onDetach()
        mUpdateStopListener = null
    }



    companion object {

        fun newInstance(): AcademicPerformanceFragment {
            return AcademicPerformanceFragment()
        }
    }
}
